pkg/peer/subscription/rtp.go

```diff
@@ -0,0 +1,151 @@
+package subscription
+
+import (
+	"fmt"
+
+	"github.com/pion/rtp"
+	"golang.org/x/exp/constraints"
+)
+
+type RewrittenRTPPacket *rtp.Packet
+
+// A structure that is used to rewrite the RTP packets that are being forwarded.
+type PacketRewriter struct {
+	// The SSRC of the previously forwarded packet.
+	previouslyForwardedSSRC uint32
+	// Currently selected SSRC of the layer (this one changes over time
+	// if we e.g. switch the layer that we want to forward).
+	selectedSSRC uint32
+	// SSRC that we're using for all packets that we're forwarding.
+	// This is the SSRC that we're sending to the remote peer. Typically,
+	// this is the SSRC of the lowest layer for the simulcast track.
+	outgoingSSRC uint32
+	// The highest identifiers of the forwarded packet,i.e. the IDs of
+	// the **latest** (in terms of timestamp and number) packet. This is not
+	// necessarily the IDs of the **last forwarded packet**, due to packets
+	// that may arrive out-of-order.
+	maxOutgoingIDs PacketIdentifiers
+	// The identifiers of the **first incoming packet** after swithing
+	// layers. We use it to calulcate their relative position in the RTP stream.
+	firstIncomingIDs PacketIdentifiers
+	// The identifiers of the **first outgoing packet** after switching
+	// layers. This is our "base" to calculate the proper timestamp when
+	// forwarding/rewriting the packet.
+	firstOutgoingIDs PacketIdentifiers
+}
+
+// Creates a new instance of the `PacketRewriter`.
+func NewPacketRewriter(outgoingSSRC uint32, selectedSSRC uint32) *PacketRewriter {
+	rewriter := new(PacketRewriter)
+	rewriter.outgoingSSRC = outgoingSSRC
+	rewriter.selectedSSRC = selectedSSRC
+	return rewriter
+}
+
+// Process new incoming packet.
+func (p *PacketRewriter) ProcessIncoming(packet *rtp.Packet) (RewrittenRTPPacket, error) {
+	// We received a packet with the SSRC different from the **currently selected** layer.
+	// This is a mistake, we drop such a package.
+	if packet.SSRC != p.selectedSSRC {
+		return nil, fmt.Errorf("Bug: packet's SSRC is different from the selected one")
+	}
+
+	// Store current incoming packet identifiers.
+	incomingIDs := PacketIdentifiers{packet.Timestamp, packet.SequenceNumber}
+
+	// Calculated outgoing IDs of the current packet.
+	outgoingIDs := PacketIdentifiers{0, 0}
+
+	// Check if we've just switched the layer before this packet, i.e. if
+	// it is the first packet after switching layers.
+	if p.previouslyForwardedSSRC != p.selectedSSRC {
+		// There is a gap of 1 seqnum to signify to the decoder that the previous frame
+		// was (probably) incomplete. That's why there's a 2 for the seqnum.
+		delta := PacketIdentifiers{1, 2}
+
+		// We make an exception for the very first packet that we're forwarding
+		// so that we start with 0 seqnum and 0 timestamp for the first packet.
+		if p.previouslyForwardedSSRC == 0 {
+			delta = PacketIdentifiers{0, 0}
+		}
```

Comment:
Can this be an if-else statement?

Reply:
I think I know what you mean :) I.e. if it was Rust I would have written it like this:

```rust
let delta = if p.previouslyForwardedSSRC == 0 {
    PacketIdentifiers::default()
} else {
    PacketIdentifiers::new(1, 2)
};
```

That's what you semantically would love to see, right?

If yes, then I'm also in favor of such initialization, but unfortunately in Go the only way to write it is either like this:

```go
var delta PacketIdentifiers
if p.previouslyForwardedSSRC == 0 {
    delta.sequenceNumber = 0, delta.timestamp = 0
} else {
    delta.sequenceNumber = 2, delta.timestamp = 1
}
```

Or, given that by default Go would initialize things with a default value (which is zero by default), shorter:

```go
var delta PacketIdenfiers
if p.previouslyForwadedSSRC != 0 {
    delta.sequenceNumber = 2, delta.timestamp = 1
}
```

The last one does not look that bad, but it seems like a variation of the current version? (except that the if is "flipped" and the initilization is default / less explicit).

Which one would you prefer?

Reply:
```go
var delta PacketIdentifiers
if p.previouslyForwardedSSRC == 0 {
    delta.sequenceNumber = 0, delta.timestamp = 0
} else {
    delta.sequenceNumber = 2, delta.timestamp = 1
}
```

^ reads the best from POV or perhaps (think that should work, right?):

```go
var delta PacketIdentifiers
if p.previouslyForwardedSSRC == 0 {
    delta = PacketIdentifiers{0, 0}
} else {
    delta = PacketIdentifiers{2, 1}
}
````

Reply:
Done.